Week_1/P1.cpp

```diff
@@ -0,0 +1,31 @@
+class Foo {
+
+private:
+    mutex gMutex;
+    condition_variable gCv;
+    int currTurn = 1;
+
+    void printsWrapper(function<void()> printFunction, int turn) {
+        unique_lock<mutex> lock(gMutex);
+        gCv.wait(lock, [this, turn]{ return turn == currTurn; });
+        printFunction();

```

Comment:
red flag - you unvoke printFunction() [on which implementation you don't have any control] under the lock.
Since you have no control on what that function does or what it can do when it is independently changed at some point in future - you can't apply "local reasoning" principle to judge about correctness of your code - and to assert correctness you would need to look at the whole program.